README.md
# GENERAL INFORMATION # 
Mode by: Ewout

Project Manager: Mr. Buys

About: Mspaint sucks so we're making our own.
 
 # SOURCES #
- JSON
- https://www.geeksforgeeks.org/read-json-file-using-python/
- https://www.w3schools.com/python/python_json.asp
# TODO #

- Make pixels change width according to chosen width
- Wrap rectangles in another list to make an easy framework for multiple frames
- 
-


- [50%COMPLETE]Recent colors
- save current color palette if new one was made
- --> save colors
- [COMPLETE]save current drawing + [PROGRESS]open a previous drawing
- -->[1/6] different ways to save to JSON (bottom to top, top to bottom, left to right, right to left, ...)
- [COMPLETE]Color palletes
- [COMPLETE]--> basic color palette at start

CHANGELOG.md
## CHANGELOG ##
- v0.3
Added function to pack an image
Added function to save pack
Made some buttons only trigger once when mouse is down instead of repeatedly.
- v0.2
Added ColorsHandler
Added recent colors, the latest 5 colors used will be saved in a list.
Made the first mode, row by row saving.
- v0.1
Forgot to keep track of all my progress but I will from now on.
Added sliders
Added custom color palettes
Added sliders for custom colors
Added dynamic pixel ranges (x,y)
Added fullscreen
Sorted everything in classes
Cleaned up some abandoned code

main.py
# 
import pygame
pygame.init()
import sys
from pygame.locals import *
import json
# ---------------------------
# CONFIGURATION VARIABLES
# ---------------------------
pixelDimentions = [16, 16]
spaceBetween = 0.1
pixel_width = 50

startingDeadSpace = 10

slider_width = 500
slider_height = 50
sliderExtraHeight = 2
slidersOffset = 30

fullscreenMode = True

colorPickerHeight = 150

customColorPalette = True
colorPaletteName = "color_palette.json"
colorPalette = 0
colorPaletteOffset = 10
colorPaletteWidth = 50

recentColorsAmount = 5

saveButtonOffsetTop = 0
saveButtonOffsetRight = 200
saveButtonWidth = 100
saveButtonHeight = 50

modeSelect = ['rij-per-rij']
mode = modeSelect[0]

extraWidth = slider_width+colorPaletteWidth*2

# ---------------------------

if customColorPalette:
    json_palette = open(colorPaletteName)
    colorPalette = json.load(json_palette)

#colours
#R    G    B
WHITE = (255, 255, 255)
BLACK = (0,   0,   0)
RED = (255, 0, 0)
GREEN = (0, 255, 0)
BLUE = (0, 0, 255)
GREY = (200, 200, 200)

BGCOLOR = BLACK

class Slider:
    def __init__(self, display, pos, width, height, color, subColor,  maximum = 255) -> None:
        self.width = width
        self.height = height
        self.pos = pos
        self.displaysurf = display
        self.rectangle = pygame.Rect(self.pos[0], self.pos[1], self.width, self.height)
        self.color = color
        self.secondaryColor = subColor
        self.max = maximum
        self.sliderCirclePosition = (self.pos[0], self.pos[1]+(self.height/2))

    def drawRect(self):
        pygame.draw.rect(self.displaysurf, self.color, self.rectangle, 2)
    
    def drawSlider(self):
        pygame.draw.circle(self.displaysurf,self.secondaryColor, self.sliderCirclePosition, self.height/2+sliderExtraHeight)

    def moveSlider(self, mousePos):
        if self.pos[0] < mousePos[0] and self.pos[1] < mousePos[1] and self.pos[0]+self.width > mousePos[0] and self.pos[1]+self.height > mousePos[1]:
            self.sliderCirclePosition = (mousePos[0], self.pos[1]+(self.height/2))

    def getColor(self):
        return round((self.sliderCirclePosition[0]/(self.pos[0]+self.width))*self.max)
            
class WindowHandler:
    def __init__(self,fullscreenMode = False) -> None:
        self.isFullscreen = fullscreenMode
        self.FPSCLOCK = pygame.time.Clock()
    
    def screenDimentions(self, pixelDimentions, pixel_width, spaceBetween = 0.5, startingDeadSpace = 20, colorPickerHeight = 150, extraWidth = 0):
        if self.isFullscreen:
            fullscreenList = pygame.display.list_modes()
            self.width = fullscreenList[0][0]
            self.height = fullscreenList[0][1]
        else:
            self.width = (pixel_width*pixelDimentions[0]) + ((pixel_width*pixelDimentions[0]) * spaceBetween) + startingDeadSpace + extraWidth
            self.height = (pixel_width*pixelDimentions[1]) + ((pixel_width*pixelDimentions[1]) * spaceBetween) + colorPickerHeight
        self.space_per_pixel_width = ((pixel_width*pixelDimentions[0]) * spaceBetween) / pixelDimentions[0]
        self.space_per_pixel_height = ((pixel_width*pixelDimentions[1]) * spaceBetween) / pixelDimentions[1]

        self.DISPLAYSURF = pygame.display.set_mode((self.width, self.height))

class Rectangle:
    def __init__(self, rectangle, startingPosition = 0, endingPosition = 0, color = WHITE) -> None:
        self.rectangle = rectangle
        self.startingPosition = startingPosition
        self.endingPosition = endingPosition
        self.color = color
    

class HandleRectangles:
    def __init__(self, pixelDimention, rectWidth) -> None:
        self.rectList = []
        self.pixelDimentions = pixelDimentions
        self.rectangleWidth = rectWidth

    def drawRectangles(self, display):
        for row in self.rectList:
            for rectangle in row:
                pygame.draw.rect(display, rectangle.color, rectangle.rectangle)

    def makeRectangleList(self, space_per_pixel_width, space_per_pixel_height):
        for index_x in range(self.pixelDimentions[0]):
            myList = []
            for index_y in range(self.pixelDimentions[1]):
                position = (space_per_pixel_width+self.rectangleWidth)*index_x+startingDeadSpace,(space_per_pixel_height+self.rectangleWidth)*index_y+startingDeadSpace
                rect = pygame.Rect(position[0], position[1],self.rectangleWidth,self.rectangleWidth)
                myList.append(Rectangle(rect, position, (position[0]+self.rectangleWidth, position[1]+self.rectangleWidth), WHITE))
            self.rectList.append(myList)
    
    def buttonHandlerForSquares(self, event):
        posInList = [0,0]
        for rectRows in self.rectList:
            for rect in rectRows:
                if rect.startingPosition[0] <= event['pos'][0] and rect.startingPosition[1] <= event['pos'][1] and rect.endingPosition[0] >= event['pos'][0] and rect.endingPosition[1] >= event['pos'][1]:
                    return rect, posInList
                posInList[1] += 1
            posInList[1] = 0
            posInList[0] += 1
        return False

class ColorPaletteHandler:
    def __init__(self, customPalette, colorPalette) -> None:
        self.colors = colorPalette
        self.customPalette = customPalette
        self.rects = []

    def rectangleConfiguration(self, pos, width):
        self.rects.append(Rectangle( pygame.Rect(pos[0]+width, pos[1], width, width), (pos[0]+width, pos[1]), (pos[0]+width+width, pos[1]+width)))
        if self.customPalette == False:
            return
        for index, color in enumerate(self.colors):
            self.rects.append(Rectangle( pygame.Rect(pos[0]+width*(index+2), pos[1], width, width), (pos[0]+width*(index+2), pos[1]), (pos[0]+width*(index+2)+width, pos[1]+width), (self.colors[color][0]*17, self.colors[color][1]*17, self.colors[color][2]*17)))
        

    def buttonHandler(self, event):
        for colorpos in self.rects:
            if colorpos.startingPosition[0] < event['pos'][0] < colorpos.endingPosition[0] and colorpos.startingPosition[1] < event['pos'][1] < colorpos.endingPosition[1]:
                return colorpos.color
            
    def drawColors(self, dislpay):
        for rect in self.rects:
            pygame.draw.rect(dislpay, rect.color, rect.rectangle)

class RecentColorsHandler:
    def __init__(self, recentColorsAmount = 5) -> None:
        self.colors = [(0, 0, 0)]
        self.maximumRecentColors = recentColorsAmount
    
    def addColor(self, color):
        if self.colors[-1] != color:
            self.colors.append(color)
        
        if len(self.colors) > self.maximumRecentColors:
            del(self.colors[0])

    def remColor(self, index):
        del(self.colors[index])

class SaveButtonHandler:
    def __init__(self, startingPos, width, height, text, mode) -> None:
        self.rectangle = Rectangle(pygame.Rect(startingPos[0], startingPos[1], width, height), startingPos, (startingPos[0]+width, startingPos[1]+height))
        self.text = text
        self.font = pygame.font.Font('freesansbold.ttf', 32)
        self.textRender = self.font.render(self.text, True, BLACK, WHITE)
        self.rectangle.rectangle = self.textRender.get_rect()
        self.rectangle.rectangle.center = ((self.rectangle.startingPosition[0]+self.rectangle.endingPosition[0])/2,(self.rectangle.startingPosition[1]+self.rectangle.endingPosition[1])/2)

        self.mode = mode

    def draw(self, display):
        display.blit(self.textRender, self.rectangle.rectangle)

    def buttonHandler(self, event):
        if self.rectangle.startingPosition[0] < event['pos'][0] < self.rectangle.endingPosition[0] and self.rectangle.startingPosition[1] < event['pos'][1] < self.rectangle.endingPosition[1]:
            return True

class MatrixData:
    def __init__(self, x_dim=0, y_dim=0, brightness=17, frame_time=0, num_frames=1,mode="rij-per-rij") -> None:
        self.x_dim = x_dim
        self.y_dim = y_dim
        self.brightness = brightness
        self.frame_time = frame_time
        self.num_frames = num_frames
        self.mode = mode
        self.frames = []

    def packFrames(self, rectList):
        self.frames = []
        if self.mode == 'rij-per-rij':
            index = 0
            while True:
                for rows in rectList:
                    self.frames.append((round(rows[index].color[0]/self.brightness), round(rows[index].color[1]/self.brightness), round(rows[index].color[2]/self.brightness)))
                    if index == len(rows)-1:
                        index = -1
                if index == -1:
                    break
                else:
                    index += 1
        print(frames)

class LoadButtonHandler:
    def __init__(self,mode) -> None:
        self.mode = mode

    def unpack(self, file):
        jsonfile = open(file, 'r')
        pixelList = []
        rowList = []

        if jsonfile['mode'] == 'rij-per-rij':
            while True:
                index = 0
                for x in range(jsonfile['x_dim']):
                    rowList.append(jsonfile['frames'][x*(index+1)])
                pixelList.append(rowList)
                if index > jsonfile['y_dim']-1:
                    return pixelList
                index += 1


def settings():
    return

def main():
    rectangleList = HandleRectangles(pixelDimentions, pixel_width)

    window = WindowHandler(fullscreenMode)
    window.screenDimentions(rectangleList.pixelDimentions, rectangleList.rectangleWidth, spaceBetween=spaceBetween, startingDeadSpace=startingDeadSpace, colorPickerHeight=colorPickerHeight, extraWidth=extraWidth)
    window.DISPLAYSURF.fill(BGCOLOR)

    rectangleList.makeRectangleList(window.space_per_pixel_width, window.space_per_pixel_height)

    redSlider = Slider(window.DISPLAYSURF, (startingDeadSpace, window.height-colorPickerHeight-30+slider_height), slider_width, slider_height/2, GREY, WHITE, 15)
    greenSlider = Slider(window.DISPLAYSURF, (startingDeadSpace, window.height-colorPickerHeight-30+slider_height + slidersOffset), slider_width, slider_height/2, GREY, WHITE, 15)
    blueSlider = Slider(window.DISPLAYSURF, (startingDeadSpace, window.height-colorPickerHeight-30+slider_height + slidersOffset*2), slider_width, slider_height/2, GREY, WHITE, 15)

    customPalette = ColorPaletteHandler( customColorPalette, colorPalette)
    customPalette.rectangleConfiguration((startingDeadSpace+slider_width+colorPaletteOffset, window.height-colorPickerHeight+slider_height), colorPaletteWidth)

    recentColors = RecentColorsHandler(recentColorsAmount)

    saveButton = SaveButtonHandler((window.width-saveButtonOffsetRight, saveButtonOffsetTop), saveButtonWidth, saveButtonHeight, "Save", mode)

    selectedColor = (0,0,0)
    mouseDown = False
    mouseDownDelay = False
    mousePos = (0,0)
    while True:

        rectangleList.drawRectangles(window.DISPLAYSURF)

        customPalette.drawColors(window.DISPLAYSURF)

        redSlider.drawRect()
        greenSlider.drawRect()
        blueSlider.drawRect()

        redSlider.drawSlider()
        greenSlider.drawSlider()
        blueSlider.drawSlider()

        saveButton.draw(window.DISPLAYSURF)

        if customPalette.rects[0].color != (redSlider.getColor() * 17, greenSlider.getColor() * 17, blueSlider.getColor() * 17):
            customPalette.rects[0].color = (redSlider.getColor() * 17, greenSlider.getColor() * 17, blueSlider.getColor() * 17)
            selectedColor = customPalette.rects[0].color
        #Good for holding down functions
        if mouseDown:
            x = rectangleList.buttonHandlerForSquares(event.dict)
            if x:
                rectangleList.rectList[x[1][0]][x[1][1]].color = selectedColor
                recentColors.addColor(selectedColor)

            redSlider.moveSlider(mousePos)
            greenSlider.moveSlider(mousePos)
            blueSlider.moveSlider(mousePos)
        #Good for once per button down functions
        if mouseDown != mouseDownDelay and mouseDown == True:
            mouseDownDelay = mouseDown

            x = saveButton.buttonHandler(event.dict)
            if x:
                new_packet = MatrixData(mode=mode)
                new_packet.packFrames(rectangleList.rectList)
                new_packet.packAll()
                new_packet.save()

            x = customPalette.buttonHandler(event.dict)
            if x:
                selectedColor = x

        for event in pygame.event.get():  # event handling loop
            if event.type == QUIT:
                terminate()
            if event.type == MOUSEBUTTONDOWN:
                mouseDown = True
            if event.type == MOUSEBUTTONUP:
                mouseDown = False
                mouseDownDelay = False
            if event.type == MOUSEMOTION:
                mousePos = event.dict['pos']

        pygame.display.update()
        window.DISPLAYSURF.fill(BGCOLOR)
        window.FPSCLOCK.tick(0)

def terminate():
    pygame.quit()
    sys.exit()

if __name__ == "__main__":
    main()

objectClasses.py
import pygame
from pygame.locals import *

class Object:
    def __init__(self, position, dimension, displaySurface, objType, color, string = '', sliderDimension = 15, sliderValueRange = [0, 15], accentColor = (0, 0, 0)) -> None:
        self.position = position
        self.dimension = dimension
        self.displaySurface = displaySurface

        self.type = objType

        self.string = string

        self.sliderValueRange = sliderValueRange
        self.sliderDimension = sliderDimension

        self.accentColor = accentColor

        self.addComponentByType()

        self.color = color

    def addAccentColor(self, color):
        self.accentColor = color

    def addCollider(self):
        self.collider = Collider(self.position, self.dimension)

    def addComponentByType(self):
        if self.type == 'slider':
            self.slider = Slider(self.position, self.dimension, self.sliderDimension, self.sliderValueRange)
            return
        
        if self.type == 'rectangle':
            self.rectangle = Rectangle(self.position, self.dimension)
            return
        
        if self.type == 'text':
            self.text = Text()
            return

    def drawObject(self):
        if self.type == 'slider':
            self.slider.updateSliderValue(self.position, self.dimension)
            pygame.draw.rect(self.displaySurface, self.color, self.slider.rectangle)
            pygame.draw.circle(self.displaySurface, self.accentColor, self.slider.center, self.slider.sliderDimension)

        if self.type == 'rectangle':
            pygame.draw.rect(self.displaySurface, self.color, self.rectangle.rectangle)

        if self.type == 'text':
            self.textFont = self.text.font.render(self.string, True, self.accentColor, self.color)
            self.displaySurface.blit(self.textFont, self.position)

class Slider:
    def __init__(self, position, dimension, sliderDimension, valueRange = [0, 15]) -> None:
        self.valueRange = valueRange
        self.currentValue = valueRange[0]
        self.center = (position[0], (position[1]+position[1]+dimension[1])/2)

        self.rectangle = pygame.Rect(position[0], position[1], dimension[0], dimension[1])
        self.sliderDimension = sliderDimension
        self.sliderPosition = (position[0], (position[1]+position[1]+dimension[1])/2)

    def updateSliderValue(self, position, dimension):
        self.currentValue = round(self.center[0]/(dimension[0]+position[0])*self.valueRange[1])

class Rectangle:
    def __init__(self, position, dimension) -> None:
        self.rectangle = pygame.Rect(position[0], position[1], dimension[0], dimension[1])


class Text:
    def __init__(self) -> None:
        self.font = pygame.font.Font('freesansbold.ttf', 32)

class Collider:
    def __init__(self, position, dimension) -> None:
        self.position = position
        self.dimension = dimension

    def checkForMouseCollision(self, event):
        if self.position[0] < event['pos'][0] < self.position[0]+self.dimension[0] and self.position[1] < event['pos'][1] < self.position[1]+self.dimension[1]:
            return True

run.py
#-------------------------
# LIBRARIES
#-------------------------
import pygame
pygame.init()
import sys
from pygame.locals import *
import json

from objectClasses import *
#-------------------------
# GLOBAL VARIABLES
#-------------------------

drawingGrid = [16, 16]

fullscreen = False
# A minimum of atleast 800x400 is required
screenResolution = [1600, 800]

screenResolutionForColors = [100, 200]

# percentage of screen what should be dead space. FOR THE WIDTH AND HEIGHT
deadSpaceBetweenGrid = 0.2

# is there a color_palette.json in the directory
isThereAColorPalette = True
colorPaletteName = 'color_palette.json'

# should remember how many colors?
recentColorAmount = 5

#RGB VALUES
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
GREY = (122, 122, 122)

colorMultiplier = 17

BGCOLOR = BLACK

#-------------------------
# BASIC CONFIGURATIONS AND CHECKS
#-------------------------

deadSpaceAtStartForRGBSliders = 20

rgbSliderDimensions = [500,20]
rgbSliderDeadSpace = 15

deadSpaceColorPalette = 20

colorPaletteWidth = 50

colorPickerDeadSpace = 10

screenResolutionForButtons = [200, 0, 10]
buttonDimensions = [100,50]

def config(screenResolution):
    minimumScreenResolution = [800, 400]
    # configuring screen height , width
    if fullscreen:
        screenResolution = pygame.display.list_modes()[0]

    elif screenResolution[0] < minimumScreenResolution[0] or screenResolution[1] < minimumScreenResolution[1]:
        print("ERR: RESOLUTION WAS NOT CONFIGURED PROPERLY, STANDARD RESOLUTION HAS BEEN USED INSTEAD.")
        screenResolution = minimumScreenResolution
    
    # configuring pixel_width and deadspace according to screen-width and height
    screenResolutionForPixels = [screenResolution[0]-screenResolutionForColors[0], screenResolution[1]-screenResolutionForColors[1]]
    pixelWidth = (screenResolutionForPixels[0]-screenResolutionForPixels[0]*deadSpaceBetweenGrid)/drawingGrid[0]
    if pixelWidth*screenResolutionForPixels[1] > screenResolution[1]:
        pixelWidth = (screenResolutionForPixels[1]-screenResolutionForPixels[1]*deadSpaceBetweenGrid)/drawingGrid[1]
    
    deadSpacePerPixel = screenResolutionForPixels[1]*deadSpaceBetweenGrid/pixelWidth

    # configuring the display surface, fps clock
    displaySurface = pygame.display.set_mode((screenResolution[0], screenResolution[1]))
    fpsClock = pygame.time.Clock()
    # return some nessesary values for other parts of the code
    return screenResolution, pixelWidth, deadSpacePerPixel, displaySurface, fpsClock

screenResolution, pixelWidth, deadSpacePerPixel, displaySurface, fpsClock = config(screenResolution)

# color palette screen resolution aftr the screenresolution was configured to avoid incorrect variables
colorPaletteDimensions = [rgbSliderDimensions[0]+deadSpaceAtStartForRGBSliders+deadSpaceColorPalette, screenResolution[1]-screenResolutionForColors[1]]

#-------------------------
# MAIN CODE
#-------------------------

def main():
    # drawing the pixel grid
    pixelGrid = []
    for x_pixel in range(drawingGrid[0]):
        y_pixels = []
        for y_pixel in range(drawingGrid[1]):
            y_pixels.append(Object(
                position= ((pixelWidth+deadSpacePerPixel)*x_pixel+deadSpacePerPixel, (pixelWidth+deadSpacePerPixel)*y_pixel+deadSpacePerPixel),
                dimension= (pixelWidth, pixelWidth),
                displaySurface= displaySurface,
                objType= 'rectangle',
                color= WHITE
            ))
            y_pixels[y_pixel].addCollider()

        pixelGrid.append(y_pixels)

    # RGB sliders
    colorSliderList = []
    for index in range(3):
        colorSliderList.append(Object(
            position= (deadSpaceAtStartForRGBSliders, (screenResolution[1]-screenResolutionForColors[1])+rgbSliderDimensions[1]*(index+1)+rgbSliderDeadSpace*index),
            dimension= rgbSliderDimensions,
            displaySurface= displaySurface,
            objType= 'slider',
            color= WHITE,
            sliderDimension= 15,
            sliderValueRange= [0, 15]
        ))
        colorSliderList[index].addAccentColor(GREY)
        colorSliderList[index].addCollider()

    # makes background of sliders Red, Green and Blue
    colorSliderList[0].color = (160, 0, 0)
    colorSliderList[1].color = (0, 160 ,0)
    colorSliderList[2].color = (0, 0, 160)

    # get the colorpalette.json imported
    colorPaletteList = [Object([colorPaletteDimensions[0], colorPaletteDimensions[1]], [colorPaletteWidth, colorPaletteWidth], displaySurface, 'rectangle', BLACK)]
    colorPaletteList[0].addCollider()
    if isThereAColorPalette:
        jsonPalette = open(colorPaletteName, 'r')
        colorPalette = json.load(jsonPalette)
        for index, color in enumerate(colorPalette):
            colorPaletteList.append(Object(
                position= [colorPaletteDimensions[0]+colorPaletteWidth*(index+1), colorPaletteDimensions[1]],
                dimension= [colorPaletteWidth, colorPaletteWidth],
                displaySurface= displaySurface,
                objType= 'rectangle',
                color= (colorPalette[color][0]*colorMultiplier, colorPalette[color][1]*colorMultiplier, colorPalette[color][2]*colorMultiplier)))
            colorPaletteList[index+1].addCollider()
    
    # recent colors configuration
    recentColors = []
    for index in range(recentColorAmount):
        recentColors.append(Object(
            position= [colorPaletteDimensions[0]+colorPaletteWidth*(index), colorPaletteDimensions[1]+colorPaletteWidth+colorPickerDeadSpace],
            dimension= [colorPaletteWidth, colorPaletteWidth],
            displaySurface= displaySurface,
            objType= 'rectangle',
            color= BLACK
        ))
        recentColors[index].addCollider()

    # load save and load buttons
    optionButtons = []
    buttons = ['save', 'load', 'settings']
    for index, string in enumerate(buttons):
        optionButtons.append(Object(
            position= (screenResolution[0]-screenResolutionForButtons[0], (screenResolutionForButtons[1]+screenResolutionForButtons[2]+buttonDimensions[1])*index),
            dimension= (buttonDimensions),
            displaySurface= displaySurface,
            objType= 'text',
            color= GREY,
            string= string
        ))
        optionButtons[index].addAccentColor(WHITE)
        optionButtons[index].addCollider()

    mouseDown = False
    mouseDownDelay = False
    selectedColor = (0,0,0)
    # main loop function
    while True:

        # draw recent colors
        for color in recentColors:
            color.drawObject()

        # draw pixel grid
        for row in pixelGrid:
            for pixel in row:
                pixel.drawObject()

        # draw sliders
        for slider in colorSliderList:
            slider.drawObject()

        # draw color palette rectangles
        for color in colorPaletteList:
            color.drawObject()

        # draw buttons
        for button in optionButtons:
            button.drawObject()

        # mouse down continuous
        if mouseDown:
            # Change selected color if sliders have been changed.
            for slider in colorSliderList:
                if slider.collider.checkForMouseCollision(event.dict):
                    slider.slider.center = (event.dict['pos'][0], slider.slider.center[1])
                    selectedColor = (colorSliderList[0].slider.currentValue * colorMultiplier, colorSliderList[1].slider.currentValue * colorMultiplier, colorSliderList[2].slider.currentValue * colorMultiplier)
                    colorPaletteList[0].color =  selectedColor

            # Change color of pixel if pixel has been pressed.
            for row in pixelGrid:
                for pixel in row:
                    if pixel.collider.checkForMouseCollision(event.dict):
                        pixel.color = selectedColor

                        # change recent color and select noew ones
                        if recentColors[0].color != selectedColor:
                            for index, color in enumerate(recentColors):
                                recentColors[len(recentColors)-index-1].color = recentColors[len(recentColors)-index-2].color
                            recentColors[0].color = selectedColor

        # mouse down once
        if mouseDown != mouseDownDelay and mouseDown == True:
            mouseDownDelay = mouseDown
            
            for color in colorPaletteList:
                if color.collider.checkForMouseCollision(event.dict):
                    selectedColor = (color.color[0], color.color[1], color.color[2])
            
            for color in recentColors:
                if color.collider.checkForMouseCollision(event.dict):
                    selectedColor = color.color

            for index, button in enumerate(optionButtons):
                if button.collider.checkForMouseCollision(event.dict):
                    if button.string == 'save':
                        print(button.string)

                    if button.string == 'load':
                        print(button.string)

                    if button.string == 'settings':
                        print(button.string)

        # event handling
        for event in pygame.event.get():
            if event.type == QUIT:
                terminate()
            if event.type == MOUSEBUTTONDOWN:
                mouseDown = True
            if event.type == MOUSEBUTTONUP:
                mouseDown = False
                mouseDownDelay = False

        # Clear the screen and prepare the following frame
        pygame.display.update()
        displaySurface.fill(BGCOLOR)
        fpsClock.tick(0)

def terminate():
    pygame.quit()
    sys.exit()

if __name__ == "__main__":
    main()